src/Mvc/Mvc.ViewFeatures.Prerendering/test/HtmlHelperComponentPrerenderingExtensionsTests.cs

```diff
@@ -0,0 +1,451 @@
+                    "attempted during prerendering or while the client is disconnected.",
+                exception.Message);
+        }
+
+        [Fact]
+        public async Task UriHelperRedirect_ThrowsInvalidOperationException()
+        {
+            // Arrange
+            var ctx = new DefaultHttpContext();
+            ctx.Request.Scheme = "http";
+            ctx.Request.Host = new HostString("localhost");
+            ctx.Request.PathBase = "/base";
+            ctx.Request.Path = "/path";
+            ctx.Request.QueryString = new QueryString("?query=value");
+
+            var helper = CreateHelper(ctx);
+            var writer = new StringWriter();
+
+            // Act
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(() => helper.PrerenderComponentAsync<RedirectComponent>(new
+            {
+                RedirectUri = "http://localhost/redirect"
+            }));
+
+            Assert.Equal("Navigation commands can not be issued at this time. This is because the component is being " +
+                    "prerendered and the page has not yet loaded in the browser or because the circuit is currently disconnected. " +
+                    "Components must wrap any navigation calls in conditional logic to ensure those navigation calls are not " +
+                    "attempted during prerendering or while the client is disconnected.",
+                exception.Message);
+        }
+
+        [Fact]
+        public async Task CanRender_AsyncComponent()
+        {
+            // Arrange
+            var helper = CreateHelper();
+            var writer = new StringWriter();
+            var expectedContent = @"<table>
+<thead>
+<tr>
+<th>Date</th>
+<th>Summary</th>
+<th>F</th>
+<th>C</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>06/05/2018</td>
+<td>Freezing</td>
+<td>33</td>
+<td>33</td>
+</tr>
+<tr>
+<td>07/05/2018</td>
+<td>Bracing</td>
+<td>57</td>
+<td>57</td>
+</tr>
+<tr>
+<td>08/05/2018</td>
+<td>Freezing</td>
+<td>9</td>
+<td>9</td>
+</tr>
+<tr>
+<td>09/05/2018</td>
+<td>Balmy</td>
+<td>4</td>
+<td>4</td>
+</tr>
+<tr>
+<td>10/05/2018</td>
+<td>Chilly</td>
+<td>29</td>
+<td>29</td>
+</tr>
+</tbody>
+</table>";
+
+            // Act
+            var result = await helper.PrerenderComponentAsync<AsyncComponent>();
+            result.WriteTo(writer, HtmlEncoder.Default);
+            var content = writer.ToString();
+
+            // Assert
+            Assert.Equal(expectedContent.Replace("\r\n", "\n"), content);
+        }
+
+
+        private class TestComponent : IComponent
+        {
+            private RenderHandle _renderHandle;
+
+            public void Configure(RenderHandle renderHandle)
+            {
+                _renderHandle = renderHandle;
+            }
+
+            public Task SetParametersAsync(ParameterCollection parameters)
+            {
+                _renderHandle.Render(builder =>
+                {
+                    var s = 0;
```

Comment:
Please remove all instances of code that generates sequence numbers programmatically. It's incredibly misleading. Users keep doing this too, perhaps because they copy examples. Sequence numbers should always be literal values.

Reply:
It's the sort of thing I'd write an analyzer for if I had enough spare time.